### PR TITLE
executor: Fix losing the auth string on changing SSL/TLS requirements

### LINKED ISSUE #25225: ALTER USER REQUIRE SSL or X509 clears out user's authentication_string (password)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

`mysql> select user,host,authentication_string from mysql.user;
+-------------+-----------+-------------------------------------------+
| user        | host      | authentication_string                     |
+-------------+-----------+-------------------------------------------+
| root        | localhost |                                           |
| backup      | %         | *9CDD29A76B21FC5186F5D9F0316EB3D7EADC695E |
| remote_root | %         | *2DBC7422585CBFB7E0404FF9E0228C33E22BD92A |
+-------------+-----------+-------------------------------------------+
3 rows in set (0.23 sec)

mysql> alter user backup require SSL;
Query OK, 0 rows affected (1.49 sec)

mysql> select user,host,authentication_string from mysql.user;
+-------------+-----------+-------------------------------------------+
| user        | host      | authentication_string                     |
+-------------+-----------+-------------------------------------------+
| root        | localhost |                                           |
| backup      | %         |                                           |
| remote_root | %         | *2DBC7422585CBFB7E0404FF9E0228C33E22BD92A |
+-------------+-----------+-------------------------------------------+
3 rows in set (0.31 sec)

mysql>`

### 2. What did you expect to see? (Required)

User require SSL or X509 to login

### 3. What did you see instead (Required)

Yes, SSL or X509 enabled. but user password also set to blank or some value, basically user can no longer login. the password must then be reset.

### 4. What is your TiDB version? (Required)

Server version: 5.7.25-TiDB-v5.0.0-rc TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #25225 

Problem Summary: An `ALTER USER ... REQUIRE SSL` would cause the authentication string of the user to get overwritten with an empty string

### What is changed and how it works?

How it Works:

- Don't change the `authentication_string` unless we have an `AuthOpt`
- Check for `SUPER` when changing other accounts auth options.

### Check List 

Tests 

- Unit test (more might be useful, but there is one already)

### Release note 

- Important security issue for handling ALTER USER statements

### CODE REVIEW

**Comment by bb7133** [comment]
Please check the failed test case(https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_check_2/detail/tidb_ghpr_check_2/9707/pipeline)
```
[2021-06-09T15:01:55.875Z] ----------------------------------------------------------------------
[2021-06-09T15:01:55.875Z] FAIL: grant_test.go:369: testSuite3.TestMaintainAuthString
[2021-06-09T15:01:55.875Z] 
[2021-06-09T15:01:55.875Z] grant_test.go:373:
[2021-06-09T15:01:55.875Z]     tk.MustExec(`CREATE USER 'maint_auth_str1'@'%' REQUIRE SSL`)
[2021-06-09T15:01:55.875Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:217:
[2021-06-09T15:01:55.875Z]     tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
[2021-06-09T15:01:55.875Z] ... value *errors.withStack = [executor:1396]Operation CREATE USER failed for 'maint_auth_str1'@'%' ("[executor:1396]Operation CREATE USER failed for 'maint_auth_str1'@'%'")
[2021-06-09T15:01:55.875Z] ... sql:CREATE USER 'maint_auth_str1'@'%' REQUIRE SSL, [], error stack [executor:1396]Operation CREATE USER failed for 'maint_auth_str1'@'%'
[2021-06-09T15:01:55.875Z] github.com/pingcap/errors.AddStack
```

**Comment by morgo** [comment]
@dveeden I've verified that it is only 4.0+ affected and removed the earlier versions from cherry-pick.

File: privilege/privileges/privileges_test.go [hunk 1]
```diff
@@ -468,6 +468,29 @@ func (s *testPrivilegeSuite) TestSetPasswdStmt(c *C) {
 	c.Assert(err, NotNil)
 }
 
+func (s *testPrivilegeSuite) TestAlterUserStmt(c *C) {
+
```
Comment (morgo) [inline privilege/privileges/privileges_test.go:472]: Please remove this blank line (style nit)

File: executor/simple.go [hunk 2]
```diff
@@ -848,10 +848,17 @@ func (e *SimpleExec) executeAlterUser(s *ast.AlterUserStmt) error {
 
 	failedUsers := make([]string, 0, len(s.Specs))
 	for _, spec := range s.Specs {
-		if spec.User.CurrentUser {
-			user := e.ctx.GetSessionVars().User
+
```
Comment (morgo) [inline executor/simple.go:851]: Also remove this line.

**Review by morgo — COMMENTED** [review]
LGTM - just some style nits.

**Comment by tiancaiamao** [comment]
/LGTM

Please address comments from @morgo

**Review by tiancaiamao — APPROVED** [review]

**Review by morgo — APPROVED** [review]
LGTM

**Review by morgo — APPROVED** [review]

**Pull request merged**